src/traces/sankey/render.js

```diff
@@ -63,44 +66,79 @@ function switchToSankeyFormat(nodes) {
-            pointNumber: i,
-            label: l,
-            color: Lib.isArrayOrTypedArray(nodeSpec.color) ? nodeSpec.color[i] : nodeSpec.color
-        };
-    });
+function sankeyModel(layout, d, traceIndex) {
+    var trace = unwrap(d).trace;
+    var domain = trace.domain;
+    var nodeSpec = trace.node;
+    var linkSpec = trace.link;
+    var arrangement = trace.arrangement;
+    var horizontal = trace.orientation === 'h';
+    var nodePad = trace.node.pad;
+    var nodeThickness = trace.node.thickness;
+    var nodeLineColor = trace.node.line.color;
+    var nodeLineWidth = trace.node.line.width;
+    var linkLineColor = trace.link.line.color;
+    var linkLineWidth = trace.link.line.width;
+    var valueFormat = trace.valueformat;
+    var valueSuffix = trace.valuesuffix;
+    var textFont = trace.textfont;
+
+    var width = layout.width * (domain.x[1] - domain.x[0]);
+    var height = layout.height * (domain.y[1] - domain.y[0]);
+
+    var links = [];
+    var hasLinkColorArray = isArrayOrTypedArray(linkSpec.color);
+    var linkedNodes = {};
+
+    var nodeCount = nodeSpec.label.length;
+    var i;
+    for(i = 0; i < linkSpec.value.length; i++) {
+        var val = linkSpec.value[i];
+        // remove negative values, but keep zeros with special treatment
+        var source = linkSpec.source[i];
+        var target = linkSpec.target[i];
+        if(!(val > 0 && isIndex(source, nodeCount) && isIndex(target, nodeCount))) {
+            continue;
+        }
+
+        source = +source;
+        target = +target;
+        linkedNodes[source] = linkedNodes[target] = true;
 
-    var links = linkSpec.value.map(function(d, i) {
-        return {
+        links.push({
             pointNumber: i,
             label: linkSpec.label[i],
-            color: Lib.isArrayOrTypedArray(linkSpec.color) ? linkSpec.color[i] : linkSpec.color,
-            source: linkSpec.source[i],
-            target: linkSpec.target[i],
-            value: d
-        };
-    });
+            color: hasLinkColorArray ? linkSpec.color[i] : linkSpec.color,
+            source: source,
+            target: target,
+            value: +val
+        });
+    }
+
+    var hasNodeColorArray = isArrayOrTypedArray(nodeSpec.color);
+    var nodes = [];
+    var removedNodes = false;
+    var nodeIndices = {};
+    for(i = 0; i < nodeCount; i++) {
+        if(linkedNodes[i]) {
+            var l = nodeSpec.label[i];
+            nodeIndices[i] = nodes.length;
+            nodes.push({
+                pointNumber: i,
+                label: l,
+                color: hasNodeColorArray ? nodeSpec.color[i] : nodeSpec.color
+            });
+        }
+        else removedNodes = true;
+    }
+
+    // need to re-index links now, since we didn't put all the nodes in
+    if(removedNodes) {
+        for(i = 0; i < links.length; i++) {
+            links[i].source = nodeIndices[links[i].source];
+            links[i].target = nodeIndices[links[i].target];
+        }
+    }
```

Comment:
For the most part it didn't cause problems to have extra unused nodes in the list, but sometimes the layout algorithm would reserve extra space for these nodes (leaving extra margins or making the whole diagram S-shaped, for example) so it's better to remove them. The `sankey_messy` mock includes a whole bunch of extra nodes both before (to test this link reassignment) and after the real ones to test this.